### PR TITLE
209: Add a check to recognize when encoding device is not present or limited (GeForce)

### LINKED ISSUE #209: Add a check to recognize when encoding device is not present
Not all Nvidia hardware support encoding. If the pipeline uses H264, HEVC, or other encoding technology that uses NVENC, we need to ensure that NVENC is present. In case it is not present like our customer encountered:

image

The system must react properly, terminating and suggesting to change output frame properties accordingly to `raw-rgba` or other in-software technology without NVENC. 

Respectively, when the system cannot allocate additional hardware ENCODER (like with GeForce), it must react properly, resulting in an understandable message and terminate clearly.

### PR DESCRIPTION
**Pull request by tomskikh**
Closes #209

### CODE REVIEW

File: gst_plugins/python/avro_video_demux.py [hunk 1]
```diff
@@ -263,14 +265,21 @@ def handle_video_frame(self, message: Dict, buffer: Gst.Buffer) -> Gst.FlowRetur
...
+                    frame = inspect.currentframe()
+                    propagate_gst_error(
+                        gst_element=self,
+                        frame=frame,
+                        file_path=__file__,
+                        domain=Gst.StreamError.quark(),
+                        code=Gst.StreamError.DEMUX,
+                        text=error,
+                    )
+                    return Gst.FlowReturn.ERROR
```
Comment (tomskikh) [inline gst_plugins/python/avro_video_demux.py:282]: Sometimes the module hangs when started simultaneously with `max_parallel_streams+1` sources.

```
2023-05-19 11:39:34,005 [ERROR] [savant.avro_video_demux] [Dummy-4] Failed to add source 'city-traffic-015ec1b103e3': reached maximum number of streams: 2.
2023-05-19 11:39:34,006 [ERROR] [savant.gstreamer.runner] [Thread-3] Received error "gst-stream-error-quark: Failed to add source 'city-traffic-015ec1b103e3': reached maximum number of streams: 2. (9)" from avro_video_demux+avrovideodemux0.
[NvMultiObjectTracker] De-initialized
```

Comment (tomskikh) [reply gst_plugins/python/avro_video_demux.py:292]: The module hangs when it tries to add elements or request pads after `shutdown()` was called. Need to check pipeline is running before adding elements or requesting pads.

File: savant/deepstream/encoding.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    pipeline: Gst.Pipeline = Gst.Pipeline.new()
+
+    converter_props = {}
+    if not is_aarch64():
+        converter_props['nvbuf-memory-type'] = int(pyds.NVBUF_MEM_CUDA_UNIFIED)
+
+    elements = [
+        PipelineElement(
+            'videotestsrc',
+            properties={'num-buffers': 50},
```
Comment (abramov-oleg) [inline savant/deepstream/encoding.py]: Do we need 50 buffers here? Maybe 1 buffer is enough?

File: savant/deepstream/encoding.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        pipeline.add(gst_element)
+        if last_gst_element is not None:
+            logger.debug('Linking %r -> %r', last_gst_element.name, gst_element.name)
+            assert last_gst_element.link(gst_element)
+        last_gst_element = gst_element
+
+    with GstPipelineRunner(pipeline) as runner:
+        while runner._is_running:
+            time.sleep(0.1)
+        if runner._error is not None:
```
Comment (abramov-oleg) [inline savant/deepstream/encoding.py:70]: Is it possible to check for a specific error message here?

File: savant/deepstream/encoding.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        else:
+            gst_element = GstElementFactory.create_element(element)
+        logger.debug('Created element %r', gst_element.name)
+        pipeline.add(gst_element)
+        if last_gst_element is not None:
+            logger.debug('Linking %r -> %r', last_gst_element.name, gst_element.name)
+            assert last_gst_element.link(gst_element)
+        last_gst_element = gst_element
+
+    with GstPipelineRunner(pipeline) as runner:
```
Comment (abramov-oleg) [inline savant/deepstream/encoding.py]: There's a new class `NvDsPipelineRunner`, can we use it here and rely on the specific error message that it builds for the case of lacking encoder sessions?

File: savant/deepstream/pipeline.py [hunk 5]
```diff
@@ -376,9 +405,11 @@ def _add_input_converter(
             source_info.before_muxer.append(video_converter)
 
         source_info.before_muxer.append(nv_video_converter)
+        self._check_pipeline_is_running()
         # TODO: send EOS to video_converter on unlink if source didn't
         assert new_pad.link(video_converter_sink) == Gst.PadLinkReturn.OK
 
+        self._check_pipeline_is_running()
```
Comment (abramov-oleg) [inline savant/deepstream/pipeline.py:412]: If I understand correctly, the method `_check_pipeline_is_running()` seems to be called repeatedly to account for possible changes to the `self._is_running` flag from parallel threads.

Isn't it better to use some synchronization mechanism to ensure immutability for this flag during the `_add_input_converter()` method (and other same casese) call?

File: savant/deepstream/encoding.py (hunk 2 above)
Comment (tomskikh) [reply savant/deepstream/encoding.py]: Yes, 1 buffer would be enough.

File: savant/deepstream/encoding.py (hunk 3 above)
Comment (tomskikh) [reply savant/deepstream/encoding.py:70]: We don't need it. The error message would be logged anyway.

File: savant/deepstream/pipeline.py (hunk 5 above)
Comment (tomskikh) [reply savant/deepstream/pipeline.py:412]: We can add a synchronization on a case when user stops the module. But when the module fails with an error synchronization would be difficult to add.

**Review by abramov-oleg — APPROVED** [review]

**Pull request merged**